### ISSUE TITLE
simple_scroll_to_fragment.html is now a perma-failure

### ISSUE THREAD

**Issue by jdm** [servo/servo#19268]
This regressed sometime on November 13, because there were no failures reported prior to that date, and since then there are >60 failures every day. This wasn't noticed because it's marked as intermittent at #14606.

**Comment by jdm** [comment]
This regression was caused by #19127. cc @tigercosmos

**Comment by jdm** [comment]
See `./mach test-wpt tests/wpt/mozilla/tests/mozilla/simple_scroll_to_fragment.html` to reproduce.

**Comment by tigercosmos** [comment]
I will check for it tomorrow.

**Comment by tigercosmos** [comment]
This is what I saw on Firefox

`simple_scroll_to_fragment.html`
![image](https://user-images.githubusercontent.com/18013815/32977815-f1f4b818-cc6f-11e7-97f3-bf5edd231ecd.png)

`simple_scroll_to_fragment_ref.html`
![image](https://user-images.githubusercontent.com/18013815/32977818-052993fe-cc70-11e7-9fa6-46b19760cbbc.png)

@jdm How do you think?

**Comment by jdm** [comment]
It does seem like the test was incorrect, but the change #19127 made no green appear at all.

**Comment by tigercosmos** [comment]
The test want to test if two show as the same, but we can see they are not equal.
Or the red area(scroll bar maybe) will not be counted in testing?

**Comment by jdm** [comment]
You are correct that the test does not pass in Firefox. However in Servo we have a layout bug that makes the test pass (which we should fix in another issue). #19127 made the test not pass in Servo, in a way that is very different from the way that it does not pass in Firefox. Does that make sense?

**Comment by tigercosmos** [comment]
I see. very weird.
Assume the wrong is right...

**Comment by tigercosmos** [comment]
what #19127 did is function `scrollTop`

`simple_scroll_to_fragment.html` has a line:
```
 document.getElementById("scroller").scrollTop = 100;
```
I would add a new test to check out whether `scrollTop` will return `100`.
If yes, that would be layout problem; if not, it is really caused by #19127

**Comment by tigercosmos** [comment]
Change to this will pass the test.
```rust
    fn has_scrolling_box(&self) -> bool {
        // TODO: scrolling mechanism, such as scrollbar (We don't have scrollbar yet)
        //       self.has_scrolling_mechanism()
        self.overflow_x_is_hidden() ||
        self.overflow_y_is_hidden() ||
        self.overflow_x_is_scroll() ||
        self.overflow_y_is_scroll()
    }

    fn has_overflow(&self) -> bool {
        self.ScrollHeight() >= self.ClientHeight() ||
        self.ScrollWidth() >= self.ClientWidth()
    }
```
Although I have find a workaround for this issue, but I think the most important reason is from #19280

**Comment by tigercosmos** [comment]
@highfive assign me

**Comment by tigercosmos** [comment]
test for this w3c/web-platform-tests#8409

**Comment by tigercosmos** [comment]
this one should be continue after #19280 done and w3c/web-platform-tests#8409 done.

**Comment by mrobinson** [comment]
Here's a related cssom-view issue: https://github.com/w3c/csswg-drafts/issues/1526

**Comment by tigercosmos** [comment]
@highfive: unassigned me.
Cannot solve this in short term.

**Comment by mrobinson** [comment]
I wonder if it makes sense to back out the original PR for the moment, since it breaks scrolling from script?

**Comment by tigercosmos** [comment]
The origin PR did fix many tests. This is the only case that failed. Beside, this issue is caused by  #19280 more precisely.

**Comment by tigercosmos** [comment]
In fact, I think `padding`, `margin`, `overflow` all have some incomplete logic, and will cause scrolling wrong, too. I believe #19466, #19477 are related to this assuming.

**Comment by mrobinson** [comment]
@tigercosmos In that case, perhaps the best thing to do here is to remove the code for "Step 10" for SetScrollTop and SetScrollLeft in order to allow scrolling from script to work, but also continuing to have the tests pass. What do you think?

**Comment by tigercosmos** [comment]
> scrolling from script

Do you mean code in [`simple_scroll_to_fragment.html`](https://github.com/servo/servo/blob/master/tests/wpt/mozilla/tests/mozilla/simple_scroll_to_fragment.html)?

In my opinion, "step 10" is followed by the SPEC to implement, and I thought this part is correct.
The key makes `SetScrollTop` wrong, which is in the test, is caused by `scrollHeight` getter, which influences "step 10". 
So, removing the code in "step 10" doesn't solve the issue, and we should fix `scrollHeight` getter first. However, `padding`, `margin`, `overflow` are incomplete currently, makes fixing `scrollHeight` more difficult.

**Comment by mrobinson** [comment]
@tigercosmos Another alternative is to remove all the new steps from SetScrollTop. In addition to the issue you mention, https://github.com/w3c/csswg-drafts/issues/1526 also causes scrolling from script to fail for nodes with overflow:scroll.

**Comment by tigercosmos** [comment]
@mrobinson I agree with you. 
`overflow:scroll` might be a potential reason why this test failed, too.
So now we have more than one factors to influence this test.


Some work we need to do:
- fix `scrollHeight`, which is a part in `has_overflow()`
- fix `has_scrolling_box()` in step 10. Should consider not only `hidden` due to  w3c/csswg-drafts#1526

https://github.com/servo/servo/blob/7a8733723551201d2c06acde9b0915b4c03938b4/components/script/dom/element.rs#L1963-L1969

**Comment by mrobinson** [comment]
@tigercosmos Since you say it is going to take a while to fix the issue, I think it makes sense to remove the code now in order to ensure that Servo can scroll from script. I can open a PR for this.

**Comment by tigercosmos** [comment]
Sure. Once test test pass, this issue will close. 
I will open a new issue to mention that some work above still need to do.

**Comment by mrobinson** [comment]
This seems to be passing in modern layout so closing for now.
